### DISCUSSION TITLE
PDM Automatic Activation Question

### DISCUSSION THREAD

**Discussion by SystemTheory** [pdm-project/pdm#1228] [Q&A] [answered]
See the tree diagram and CLI commands below. The `(fastapi-env)` is automatically activated when I change into any directory under pyenv-test-env and deactivated when I change to another directory. This is controlled by a line inserted in the `.python-version` file. The packages `fastapi` and `uvicorn` are installed in the pyenv virtualenv using pip-tools. If I run `uvicorn main:app` under the directory `0000-hello-fastapi` this test works fine.

I had hoped to setup PDM via venv and pep582 with the same ability to activate and deactivate. Either I am not applying the PDM documentation properly or pyenv virtualenv is interfering with PDM.

Can I get PDM to work in this context alonside pyenv virtualenv tools?

If possible how do I configure PDM to work like pyenv virtualenv that activates the environment or finds the pep582 packages automatically when inside the project folder and deactivates when outside the folder?

```
$ tree ~/pytest -aL 2
/home/joe/pytest
├── pdm-test-pep582
│   ├── 0000-hello-fastapi
│   ├── pdm.lock
│   ├── .pdm.toml
│   ├── __pypackages__
│   └── pyproject.toml
├── pdm-test-venv
│   ├── 0000-hello-fastapi
│   ├── pdm.lock
│   ├── .pdm.toml
│   ├── pyproject.toml
│   └── .venv
└── pyenv-test-venv
    ├── 0000-hello-fastapi
    ├── .python-version
    ├── requirements.in
    └── requirements.txt

/home/joe/pytest/pyenv-test-venv/0000-hello-fastapi
$ echo; cat main.py; echo

from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}

(fastapi-env)

/home/joe
$ pdm -V
PDM, version 2.0.0b2

/home/joe/pytest/pyenv-test-venv/0000-hello-fastapi
$ cd ~/pytest/pdm*venv/0000-hello-fastapi

/home/joe/pytest/pdm-test-venv/0000-hello-fastapi
$ eval $(pdm venv activate in-project)
(.venv)

/home/joe/pytest/pdm-test-venv/0000-hello-fastapi
$ uvicorn main:app
pyenv: uvicorn: command not found

The `uvicorn' command exists in these Python versions:
  3.10.4/envs/fastapi-env
  fastapi-env

Note: See 'pyenv help global' for tips on allowing both
      python2 and python3 to be found.
(.venv)

```

**Comment by frostming** [comment, marked as answer]
> If possible how do I configure PDM to work like pyenv virtualenv that activates the environment or finds the pep582 packages automatically when inside the project folder and deactivates when outside the folder?

If you mean to auto-activate the **shell** so that executables inside the venv are brought to the front in the `PATH`, PDM doesn't provide such ability. You can still create venv with `pyenv` and use it to activate for you. PDM will respect the `VIRTUAL_ENV` environment variable. Just make sure you don't store the python path in `.pdm.toml`, to delete it: `pdm config -ld python.path`.

**Reply by SystemTheory** [reply]
In my user account on WSL2 Debian I will probably use `pyenv` to manage python versions; `direnv` to install and manage virtual environments using the direnv command `layout pyenv 3.10.4` (for example); and `pip-tools` with `pip` to manage packages in each environment. The links below provide more information on this workflow in case others are interested:

https://stackabuse.com/managing-python-environments-with-direnv-and-pyenv/

https://pwal.ch/posts/2019-11-10-sane-python-environment-2020-part-4-virtualenv/

Lonesnake is another option for making self-contained python versions and virtual environments in the project folder:

https://github.com/pwalch/lonesnake

Although I like some of the features of PDM my goal is to see only the list of installed python versions when I type `pyenv versions`; to put the python virtual environment in the project folder; to activate the virtual environment inside the project folder automatically; and to link the python version explicitly to the virtual environment with a simple commandline directive. I can do this with `pyenv` and `direnv`.
